### ISSUE TITLE
twisted.test.test_paths.FilePathTestCase.test_getPermissions_Windows creates undeleteable file

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5228]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5228|
|Type|defect|
|Created|2011-08-09 16:22:34Z|

This test method sets the mode of sub1 such that it cannot be deleted in the usual way:

    [Error 5] Access is denied: '_trial_temp\\twisted.test.test_paths\\FilePathTestCase\\test_getPermissions_Windows\\bvk9lu\\temp\\sub1'

The test should ensure that regardless of the test outcome, this file ends up deletable, or it should delete it itself.

Attachments:

* [changed_file_permission_to_777.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fad/fad9483d061de3d1f1c61377eb42c51a3c38b082/c85d7c52aec3c9df4ed3823336807c1d8c2ac916.patch) (478 bytes) - added by dhanush on 2013-03-31 12:04:46Z - Changed the file permission to 777 in lieu of the sanity check.
* [create_deleteable_files.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fad/fad9483d061de3d1f1c61377eb42c51a3c38b082/ad031b341387327893a8d3da134977011fd73cfb.patch) (807 bytes) - added by mhaarhaus on 2013-04-15 07:02:02Z - Change the final chmod command in test_getPermissions_Windows to allow sub1 to be deleted, while maintaining the user permissions = group permissions = other permissions test.
* [change_permissions_after_test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fad/fad9483d061de3d1f1c61377eb42c51a3c38b082/042a81d4223ec84d0d7cb428b79ab9bd0b00c99a.patch) (581 bytes) - added by mhaarhaus on 2013-04-16 01:15:44Z - Change the file permissions during cleanup to allow for deletion.
* [change_permissions_after_test2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fad/fad9483d061de3d1f1c61377eb42c51a3c38b082/a2d51615f59e18c77978284aee21639ac23bfe5e.patch) (918 bytes) - added by mhaarhaus on 2013-04-16 17:13:49Z - Include the news file.
* [5228.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fad/fad9483d061de3d1f1c61377eb42c51a3c38b082/2da412c5e47b3ef4b20d9641139caa11ebb65d32.patch) (715 bytes) - added by mhaarhaus on 2013-04-29 06:55:14Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5228 5228
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1312906954000000 1312906954000000
changetime__1367236498000000 1367236498000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="gdeng's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|gdeng set owner to gdeng|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="dhanush's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|dhanush set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to dhanush|
|-|-|

Thanks for working on this dhanush.  I'm not sure why this is assigned to me, though.  The general procedure is to assign tickets for review to noone (the empty entry at the top of the "assign to" dropdown).  I filed this ticket, but that doesn't mean I have to review it.

However, since I'm here and the patch is very short, I guess I will.

My only comment on the patch is that it doesn't seem to address the issue.  It changes an assertion being made in the test method.  This doesn't seem correct to me, the assertion being made before was correct, so the new assertion after this patch is probably incorrect.    The tests only runs on Windows so I can't easily run it, but I suppose that if it were run, it would fail.

Beyond that, the problem described in the ticket description still exists.  The file is still created with permissions that make it undeleteable.

**Comment by twisted-trac** [comment]
|<img alt="dhanush's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|dhanush commented|
|-|-|

Thanks for your feedback. I will stick to the general procedure for assigning tickets to review. Regarding the patch, as the issue still stands, I'll try to take a plunge in more detail into the code and try to fix this up in a future patch.

**Comment by twisted-trac** [comment]
|<img alt="mhaarhaus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|mhaarhaus removed owner|
|-|-|

Change the final chmod command in test_getPermissions_Windows to allow sub1 to be deleted, while maintaining the user permissions = group permissions = other permissions test.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince set owner to mhaarhaus|
|-|-|

The change looks good, but it could probably be better. If the assertion triggers the second time through the loop, it will still leave an undeletable file.

[TestCase.addCleanup](https://twistedmatrix.com/documents/current/api/twisted.trial.unittest.TestCase.html#addCleanup) could be used to avoid this. It would also be useful to add a comment as to why it needs to be done, so future maintainers don't break it again.

**Comment by twisted-trac** [comment]
|<img alt="mhaarhaus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|mhaarhaus commented|
|-|-|

Maybe I'm looking at it wrong, but isn't the final chmod (to 511, making it undeleteable) occur outside of the loop? The mode is changed to 777 and tested then 555 and tested inside the loop, then with my change, 711 and tested outside.

Regardless, I'll look into doing it with addCleanup instead.

**Comment by twisted-trac** [comment]
|<img alt="mhaarhaus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|mhaarhaus removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_9"></a>|@tomprince commented|
|-|-|

> Maybe I'm looking at it wrong, but isn't the final chmod (to 511, making it undeleteable) occur outside of the loop? The mode is changed to 777 and tested then 555 and tested inside the loop, then with my change, 711 and tested outside.

If none of the assertions fail, then that is the case. But if one the assertions fails, then the code will interrupted, and the file could be left in an undeletable state.

**Comment by twisted-trac** [comment]
|<img alt="mhaarhaus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|mhaarhaus commented|
|-|-|

Replying to [tom.prince](#note_9):
> > Maybe I'm looking at it wrong, but isn't the final chmod (to 511, making it undeleteable) occur outside of the loop? The mode is changed to 777 and tested then 555 and tested inside the loop, then with my change, 711 and tested outside.
> 
> If none of the assertions fail, then that is the case. But if one the assertions fails, then the code will interrupted, and the file could be left in an undeletable state.

Okay, I understand. The new patch uses addCleanup, so it should avoid that issue.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst set owner to mhaarhaus|
|-|-|

The problem with doing the cleanup registration at the end of the function is that if one of the assertion fails earlier one, or the test is interrupted, it will never be reached. So cleanup will still not happen if an assertion fails early on.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_12"></a>|@itamarst commented|
|-|-|

Oh, and the change is minor enough the news file could just be a .misc with no content.

**Comment by twisted-trac** [comment]
|<img alt="mhaarhaus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|mhaarhaus removed owner|
|-|-|

Replying to [itamar](#note_11):
> The problem with doing the cleanup registration at the end of the function is that if one of the assertion fails earlier one, or the test is interrupted, it will never be reached. So cleanup will still not happen if an assertion fails early on.

Yeah, silly mistake. Fixed that and changed the news file.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks!  This looks like the right fix.  I'll apply it to trunk.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [38294]) Clean up test state in the `FilePath` tests on Windows so that `_trial_temp` can be deleted

Author: mhaarhaus
Reviewer: tom.prince, itamar, exarkun
Fixes: [#5228](https://github.com/twisted/twisted/issues/5228)

Previously a `FilePath` test left some files in `_trial_temp` without permissions
necessary to delete them.  This made clean up _trial_temp a pain.  The test now
makes sure to restore permissions on the dummy files after the test is over.
